


##################
# Input data
##################
# Directory containing cafeh scripts
cafeh_code_dir="/work-zfs/abattle4/bstrober/tools/cafeh/"

# Directory containing gtex cafeh data (generated by karl)
gtex_cafeh_data="/work-zfs/abattle4/karl/cosie_analysis/output/GTEx/"

# Directory containing summary statistics
# contains file called study_info explaining files and where they came from
summary_stat_dir="/work-zfs/abattle4/bstrober/eqtl_informed_prs/input_data/sumstats/"


# File containing list of gene-names and whether or not they are protein-coding
gene_info_file="/work-zfs/abattle4/lab_data/GTEx_v8/references/gencode.v26.GRCh38.genes.gtf"

#Directory that contains necessary liftover information.
##Specifically, it must contain:
#####1. 'liftOver'   --> the executable
#####2. 'hg19ToHg38.over.chain.gz'   --> for converting from hg19 to hg38
#####2. 'hg38ToHg19.over.chain.gz'   --> for converting from hg38 to hg19
liftover_directory="/work-zfs/abattle4/bstrober/tools/liftOver_x86/"

# File containing gtex tissues to do analysis on and their sample size
gtex_tissue_file="/work-zfs/abattle4/bstrober/eqtl_informed_prs/input_data/gtex_tissue_sample_size_subset.txt"


##################
# Output data
##################
# Output root directory
output_root="/work-zfs/abattle4/bstrober/eqtl_informed_prs/gtex_v8/"

# Directory containing all gtex associations (in cafeh format)
processed_gtex_associations_dir=$output_root"processed_gtex_associations/"

# Directory containing all gtex associations (in cafeh format)
processed_bivariate_cafeh_input_dir=$output_root"processed_bivariate_cafeh_input/"

# Directory containing CAFEH results
bivariate_cafeh_output_dir=$output_root"bivariate_cafeh_output/"





if false; then
sh process_gtex_associations_for_cafeh.sh $processed_gtex_associations_dir $gene_info_file $liftover_directory
fi

trait_name="blood_white_count"
trait_sumstat_file=$summary_stat_dir"bolt_337K_unrelStringentBrit_MAF0.001_v3.blood_WHITE_COUNT.bgen.stats.gz"
sample_size="337491"

trait_name="whr_adjusted_bmi"
trait_sumstat_file=$summary_stat_dir"bolt_337K_unrelStringentBrit_MAF0.001_v3.body_WHRadjBMIz.bgen.stats.gz"
sample_size="487409"
sh process_data_for_bivariate_cafeh_analysis.sh $trait_name $trait_sumstat_file $gtex_tissue_file $gtex_cafeh_data $processed_gtex_associations_dir $processed_bivariate_cafeh_input_dir $sample_size


if false; then
sh run_cafeh_on_bivariate_data_for_single_trait_shell.sh $trait_name $gtex_tissue_file $processed_bivariate_cafeh_input_dir $bivariate_cafeh_output_dir
fi















# CURRENTLY HACKY, switching over to o2

##################
# Input data
##################
# UKBB download data
ukbb_download_data="/n/groups/price/UKBiobank/download_500K/"
# UKBB sumstats
ukbb_sumstats_dir="/n/groups/price/UKBiobank/sumstats/bolt_337K_unrelStringentBrit_MAF0.001_v3/"
# Beta files
cafeh_prs_betas_dir="/n/groups/price/ben/eqtl_informed_prs/input_data/"
# File containing estimates of number of cafeh components per tissue
num_components_per_tissue_file="/n/groups/price/ben/eqtl_informed_prs/input_data/cafeh_results_blood_white_count_num_prs_components.txt"
# UKBB Phenotype files
ukbb_pheno_file1="/n/groups/price/steven/RareVariants/Final/UKB_new_sumstats/UKB_v3.061518.tab"
ukbb_pheno_file2="/n/groups/price/UKBiobank/app19808mosaic/bloodQC/ukb4777.blood_v2.covars.tab"
ukbb_pheno_file3="/n/groups/price/UKBiobank/app10438assoc/ukb4777.processed_and_post.plinkPCs.tab.gz"



##################
# Output data
##################
# Root output directory
output_root="/n/groups/price/ben/eqtl_informed_prs/"
# Directory containing filtered UKBB sample names
ukbb_sample_names_dir=$output_root"ukbb_sample_names/"
# Directory containing generate PRS
ukbb_prs_dir=$output_root"ukbb_prs/old/"
ukbb_prs_dir=$output_root"ukbb_prs/"
# Directory containing generate PRS
ukbb_prs_viz_dir=$output_root"visualize_ukbb_prs/"

trait_name="blood_white_count"



if false; then
sh generate_ukbb_sample_name_lists.sh $ukbb_download_data $ukbb_sample_names_dir
fi



beta_thresholds=( "1" ".5" ".1" ".05" ".01" ".005")
if false; then
for beta_threshold in "${beta_thresholds[@]}"; do	
	for chrom_num in $(seq 1 22); do
		sbatch generate_prs.sh $ukbb_sample_names_dir $ukbb_download_data $cafeh_prs_betas_dir $ukbb_prs_dir $chrom_num $beta_threshold
	done
done
fi

if false; then
sh organize_prs_results.sh $ukbb_prs_dir $ukbb_pheno_file1 $ukbb_pheno_file2 $ukbb_pheno_file3
fi


if false; then
module load R/3.5.1
Rscript visualize_prs_results.R $ukbb_prs_dir $ukbb_prs_viz_dir $num_components_per_tissue_file
fi






















#################
#### OLD
##################
# Directory containing processed UKBB genotype data
processed_ukbb_genotype_dir=$output_root"processed_ukbb_genotype/"


if false; then
sh process_ukbb_genotype_data.sh $ukbb_sample_names_dir $ukbb_download_data $ukbb_sumstats_dir $processed_ukbb_genotype_dir
fi
