import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from itertools import islice
import glob
import json
import pdb
import sys
import os


#########################
# Code generated by Karl Tayeb (with minor adjustments)
# https://github.com/karltayeb/coloc-finemap/blob/master/workflow/scripts/get_gtex_associations.py
##########################

def get_gene_id_and_type(gene_info):
    gene_id = ''
    gene_type = ''
    for ele in gene_info:
        if ele.split(' "')[0] == 'gene_id':
            if gene_id != '':
                print('assumption eroror')
                pdb.set_trace()
            gene_id = ele.split(' "')[1].split('"')[0]
        elif ele.split(' "')[0] == ' gene_type':
            if gene_type != '':
                print('assumtpione oeorroro')
                pdb.set_trace()
            gene_type = ele.split(' "')[1].split('"')[0]
    if gene_id == '' or gene_type == '':
        print('never made it')
        pdb.set_trace()
    return gene_id, gene_type

def get_protein_coding_autosomal_genes(gene_info_file):
    valid_chromosomes = {}
    for chrom_num in range(1,23):
        chrom_string = 'chr' + str(chrom_num)
        valid_chromosomes[chrom_string] = 1

    f = open(gene_info_file)
    pc_genes = {}
    for line in f:
        line = line.rstrip()
        data = line.split('\t')
        if line.startswith('##'):
            continue
        # Remove non-autosomal genes
        if data[0] not in valid_chromosomes:
            continue
        gene_info = data[8].split(';')
        gene_id, gene_type = get_gene_id_and_type(gene_info)
        if gene_type == 'protein_coding':
            pc_genes[gene_id] = 1
    f.close()
    return pc_genes

def get_karls_genes(dir_root):
    karls_genes = {}
    for chrom_num in range(1,23):
        chrom_root = dir_root + 'chr' + str(chrom_num) + '/'
        for dir_name in os.listdir(chrom_root):
            if dir_name.startswith('ENSG') == False:
                continue
            if dir_name.endswith('.tar.gz'):
                continue
            gene_dir = chrom_root + dir_name + '/'
            variant_report_file = gene_dir + dir_name + '.cafeh_genotype_ss.variant_report'
            if os.path.exists(variant_report_file):
                karls_genes[dir_name] = 1
    return karls_genes


cafeh_gene_list = sys.argv[1]
processed_gtex_associations_dir = sys.argv[2]
job_number = int(sys.argv[3])
total_jobs = int(sys.argv[4])

association_indices = {
        x.split('/')[-1].split('.')[0]: json.load(open(x, 'r'))
        for x in glob.glob('/work-zfs/abattle4/karl/cosie_analysis/output/GTEx/index/*.index')
    }



gene_tissue_map = {}
for tissue in association_indices:
    for gene in association_indices[tissue]:
        if gene == 'gene_id':
            continue
        if gene in gene_tissue_map:
            gene_tissue_map[gene].append(tissue)
        else:
            gene_tissue_map[gene] = [tissue]

def get_associations(test_gene):
    results = {}
    for test_tissue in gene_tissue_map[test_gene]:
        lines = []
        with open(
            '/work-zfs/abattle4/lab_data/GTEx_v8/ciseQTL/'
            'GTEx_Analysis_v8_eQTL_all_associations/{}.allpairs.txt'.format(
                test_tissue), 'r') as f:
            f.seek(association_indices[test_tissue][test_gene])
            last_gene = ''
            for line in f:
                gene = line.split('\t')[0]
                if gene != test_gene and len(lines) > 1:
                    break
                else:
                    lines.append(line)
                    last_gene = gene
        results[test_tissue] = pd.DataFrame([x.strip().split('\t') for x in lines])

    results = pd.concat(results)
    results = results.rename({
        0: 'gene_id',
        1: 'variant_id',
        2: 'tss_distance',
        3: 'ma_samples',
        4: 'ma_count',
        5: 'maf',
        6: 'pval_nominal',
        7: 'slope',
        8: 'slope_se'
    }, axis='columns')

    results.slope = results.slope.astype(float)
    results.slope_se = results.slope_se.astype(float)

    results = results[results.maf.astype(float) > 0.01]
    results.index.levels[0].name = 'tissue'
    results.reset_index(inplace=True)
    results.drop('level_1', axis=1, inplace=True)
    return results



def get_number_of_genes(cafeh_gene_list):
    f = open(cafeh_gene_list)
    counter = 0
    head_count = 0
    genes = []
    chroms = []
    for line in f:
        line = line.rstrip()
        data = line.split('\t')
        counter = counter + 1
        if head_count == 0:
            head_count = head_count + 1
            continue
        genes.append(data[0])
        chroms.append(data[1])
    f.close()
    if counter -1 != len(genes):
        print('assumptioneorororo')
        pdb.set_trace()
    return counter -1, np.asarray(genes), np.asarray(chroms)


# For parallelization purposes
def parallelization_start_and_end(num_tasks, job_number, total_jobs):
    tasks_per_job = (num_tasks/total_jobs) + 1
    start_task = job_number*tasks_per_job
    end_task = (job_number + 1)*tasks_per_job -1 
    return start_task, end_task


def get_list_of_genotyped_variants(variant_file):
    f = open(variant_file)
    dicti = {}
    head_count = 0
    for line in f:
        line = line.rstrip()
        data = line.split()
        if head_count == 0:
            head_count = head_count + 1
            variants = data[6:]
            for variant in variants:
                variant = '_'.join(variant.split('_')[:5])
                dicti[variant] = 1
            continue
        break
    f.close()
    return dicti


num_genes, genes, chroms = get_number_of_genes(cafeh_gene_list)


#For parallelization purposes
start_number, end_number = parallelization_start_and_end(num_genes, job_number, total_jobs)

for gene_index, test_gene in enumerate(genes):
    # Skip genes not in this parallelization run
    if gene_index < start_number or gene_index > end_number:
        continue
    chrom = chroms[gene_index]
    # file containing variants for this gene
    variant_file = '/work-zfs/abattle4/karl/cosie_analysis/output/GTEx/' + chrom + '/' + test_gene + '/' + test_gene + '.raw'
    genotyped_variant_list = get_list_of_genotyped_variants(variant_file)
    results = get_associations(test_gene)
    # Subset to snps we have genotype data for (useful in computing r-squared)
    boolean_arr = []
    for snp_id in results['variant_id']:
        if snp_id in genotyped_variant_list:
            boolean_arr.append(True)
        else:
            boolean_arr.append(False)
    boolean_arr = np.asarray(boolean_arr)
    results = results[boolean_arr]
    # Quick error check
    for snp_id in results['variant_id']:
        if snp_id not in genotyped_variant_list:
            print('assumption eroror')
            pdb.set_trace()
    output_file = processed_gtex_associations_dir + test_gene + '_associations.csv'
    results.to_csv(output_file)

